src/Native/Unix/System.Security.Cryptography.Native/pal_x509.c

```diff
@@ -323,3 +387,659 @@ X509* CryptoNative_X509UpRef(X509* x509)
+
+        // Treat "response has no nonce" as success, since not all responders set the nonce.
+        if (nonceCheck == -1)
+        {
+            nonceCheck = 1;
+        }
+
+        if (nonceCheck == 1 && OCSP_basic_verify(basicResp, untrusted, store, 0))
+        {
+            ASN1_GENERALIZEDTIME* thisupd = NULL;
+            ASN1_GENERALIZEDTIME* nextupd = NULL;
+
+            if (OCSP_resp_find_status(basicResp, certId, &status, NULL, NULL, &thisupd, &nextupd))
+            {
+                if (thisUpdate != NULL && thisupd != NULL)
+                {
+                    *thisUpdate = ASN1_STRING_dup(thisupd);
+                }
+
+                if (nextUpdate != NULL && nextupd != NULL)
+                {
+                    *nextUpdate = ASN1_STRING_dup(nextupd);
+                }
+
+                if (status == V_OCSP_CERTSTATUS_GOOD)
+                {
+                    ret = PAL_X509_V_OK;
+                }
+                else if (status == V_OCSP_CERTSTATUS_REVOKED)
+                {
+                    ret = PAL_X509_V_ERR_CERT_REVOKED;
+                }
+            }
+        }
+
+        OCSP_BASICRESP_free(basicResp);
+        basicResp = NULL;
+    }
+
+    OCSP_CERTID_free(certId);
+    return ret;
+}
+
+static int Get0CertAndIssuer(X509_STORE_CTX* storeCtx, X509** subject, X509** issuer)
+{
+    assert(storeCtx != NULL);
+    assert(subject != NULL);
+    assert(issuer != NULL);
+
+    // get0 => don't free.
+    X509Stack* chain = X509_STORE_CTX_get0_chain(storeCtx);
+    int chainSize = chain == NULL ? 0 : sk_X509_num(chain);
+
+    if (chainSize < 1)
+    {
+        return 0;
+    }
+
+    *subject = sk_X509_value(chain, 0);
+    *issuer = sk_X509_value(chain, chainSize == 1 ? 0 : 1);
+    return 1;
+}
+
+static time_t GetIssuanceWindowStart()
+{
+    // time_t granularity is seconds, so subtract 4 days worth of seconds.
+    // The 4 day policy is based on the CA/Browser Forum Baseline Requirements
+    // (version 1.6.3) section 4.9.10 (On-Line Revocation Checking Requirements)
+    time_t t = time(NULL);
+    t -= 4 * 24 * 60 * 60;
+    return t;
+}
+
+X509VerifyStatusCode CryptoNative_X509ChainGetCachedOcspStatus(X509_STORE_CTX* storeCtx, char* cachePath)
+{
+    if (storeCtx == NULL || cachePath == NULL)
+    {
+        return (X509VerifyStatusCode)-1;
+    }
+
+    X509* subject;
+    X509* issuer;
+
+    if (!Get0CertAndIssuer(storeCtx, &subject, &issuer))
+    {
+        return (X509VerifyStatusCode)-2;
+    }
+
+    X509VerifyStatusCode ret = PAL_X509_V_ERR_UNABLE_TO_GET_CRL;
+    char* fullPath = BuildOcspCacheFilename(cachePath, subject);
+
+    if (fullPath == NULL)
+    {
+        return ret;
+    }
+
+    BIO* bio = BIO_new_file(fullPath, "rb");
+    OCSP_RESPONSE* resp = NULL;
+
+    if (bio != NULL)
+    {
+        resp = d2i_OCSP_RESPONSE_bio(bio, NULL);
+        BIO_free(bio);
+    }
+
+    if (resp != NULL)
```

Comment:
This could all be part of the if bio != NULL, right?

Reply:
The expected state is that everything succeeds until possibly the time checks both `if`s are "always" evaluated, and not nesting reduces indentation.

But I can nest it if you feel like it's not scanning well.

Reply:
Ok.